### PR TITLE
SLING-11450 : avoid suppressing instances that were seen previously

### PR DESCRIPTION
**Pull request by stefan-egli**
in the local cluster, prior to a topology change. With this, the partial-startup-suppression / joinerdelay feature again respects leaders as required by the API

### CODE REVIEW

**Comment by mreutegg** [comment]
I can reproduce those test failures reported by Jenkins and did some debugging to find out why they fail. I may not have an in-depth understanding of the implementation, but one area that looks weird to me is how IdMapService reads the idMap.

E.g. when I debug PartialStartupTest.testNoLocalSyncTokenYet(), then I can see that at some point IdMapService.readIdMap() reads 4 id to slingId mappings from /var/discovery/oak/idMap. However, the returned Map only contains 3 entries, because on the node in the repository there are two slingId properties that point to the same id.

- 289b8024-1437-4f5a-984c-97d01f5a5456 -> 2
- 0384e415-bd6f-477b-b32d-4387b06330ee -> 1
- 3fce3eb8-3a7e-423a-b662-f1880cb75e49 -> 3
- 226d18c3-d979-42c6-ab1a-f1c641dd8c31 -> 3

Doesn't this cause problems later? It appears this can randomly hide some mapping.

**Comment by stefan-egli** [comment]
This is due to a bug in discovery.base which the other PR https://github.com/apache/sling-org-apache-sling-discovery-base/pull/9 fixes. I should have noted this in the PR, my bad. (and for some reason I missed the jenkins warnings)

File: src/main/java/org/apache/sling/discovery/oak/cluster/ClusterReader.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        }
+        final String syncTokenStr = resourceMap.get(slingId, String.class);
+        if (syncTokenStr == null) {
+            return -1;
+        }
+        try {
+            return Long.parseLong(syncTokenStr);
+        } catch (NumberFormatException nfe) {
+            logger.warn(
+                    "readSyncToken: unparsable (non long) syncToken: " + syncTokenStr);
```
Comment (mreutegg) [inline src/main/java/org/apache/sling/discovery/oak/cluster/ClusterReader.java:79]: String concatenation can be avoided with format specifier.
```suggestion
                    "readSyncToken: unparsable (non long) syncToken: {}", syncTokenStr);
```

File: src/main/java/org/apache/sling/discovery/oak/cluster/ClusterReader.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        }
+        final boolean hasSeenLocalInstance = hasSeenLocalinstance(clusterNodeId, slingId, leaderElectionId);
+        final long syncToken = readSyncToken(slingId);
+        if (syncToken == -1 && failOnMissingSyncToken && !hasSeenLocalInstance) {
+            return InstanceReadResult.fromErrorMsg("no syncToken available yet for slingId=" + slingId);
+        }
+        return InstanceReadResult.fromInstance(new InstanceInfo(clusterNodeId, slingId, syncToken, leaderElectionId));
+    }
+
+    private boolean hasSeenLocalinstance(int clusterNodeId, String slingId,
```
Comment (mreutegg) [inline src/main/java/org/apache/sling/discovery/oak/cluster/ClusterReader.java:133]: Minor typo. I would rather write `hasSeenLocalInstance` instead. Or did you deliberately use a lower case `i` to better distinguish it from the preceding `l`?

File: src/main/java/org/apache/sling/discovery/oak/cluster/OakClusterViewService.java [hunk 3]
```diff
@@ -181,74 +198,108 @@ private LocalClusterView asClusterView(DiscoveryLiteDescriptor descriptor, Resou
...
-            if (slingId==null) {
+            InstanceInfo in = regularInstances.get(id);
+            String slingId = in == null ? null : in.getSlingId();
+            if (slingId == null) {
                 idMapService.clearCache();
                 logger.info("asClusterView: cannot resolve oak-clusterNodeId {} to a slingId", id);
                 throw new Exception("Cannot resolve oak-clusterNodeId "+id+" to a slingId");
             }
+            if (!in.isSyncTokenNewerOrEqual(seqNum)) {
+                System.out.println("Not seen syncToken (" + seqNum + ") of this instance yet : " + in);
```
[the remark this answers is not in the document]
Comment (stefan-egli) [reply src/main/java/org/apache/sling/discovery/oak/cluster/OakClusterViewService.java]: oups, fixed.

File: src/test/java/org/apache/sling/discovery/oak/its/setup/SimulatedLeaseCollection.java [hunk 4]
```diff
@@ -114,6 +123,17 @@ private int getClusterNodeId(SimulatedLease lease) {
         return getClusterNodeId(lease.getSlingId(), lease.getClusterNodeIdHint());
     }
 
+    int getClusterNodeId(String slingId) {
+        if (clusterNodeIds == null) {
+            return -1;
+        }
```
Comment (mreutegg) [inline src/test/java/org/apache/sling/discovery/oak/its/setup/SimulatedLeaseCollection.java:129]: `clusterNodeIds` is never null. Remove those lines.
```suggestion
```

File: src/test/java/org/apache/sling/discovery/oak/its/setup/SimulatedLeaseCollection.java [hunk 5]
```diff
@@ -70,6 +70,15 @@ public void incSeqNum(int amount) {
 
     public synchronized void hooked(SimulatedLease lease) {
         leases.add(lease);
+//        if (leases.size() > 1) {
+//            incSeqNum();
+//        }
```
Comment (mreutegg) [inline src/test/java/org/apache/sling/discovery/oak/its/setup/SimulatedLeaseCollection.java:75]: Remove?
```suggestion
```

**Comment by mreutegg** [comment]
Generally looks good, except for some minor things I commented on earlier.

File: src/test/java/org/apache/sling/discovery/oak/its/setup/SimulatedLeaseCollection.java (hunk 5 above)
Comment (stefan-egli) [reply src/test/java/org/apache/sling/discovery/oak/its/setup/SimulatedLeaseCollection.java:75]: yes, should remove - left over from testing

**Comment by stefan-egli** [comment]
jenkins builds fine now.
@mreutegg , I sort of read your comment as an official +1? :)

**Review by mreutegg — APPROVED** [review]
Yes, looks good to me.

**Pull request merged**